### ISSUE TITLE
Use everywhere replication strategy for system_auth

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#1796]
*Installation details*
Scylla version (or git commit hash): 1.4

Rather than updating the replication factor all the time for `system_auth`, use the everywhere replication strategy.

**Comment by slivne** [comment]
https://issues.apache.org/jira/browse/CASSANDRA-12629

there are reasons not todo this - lets review the cons as well.g

**Comment by slivne** [comment]
@tzach any ideas ?

**Comment by penberg** [comment]
The Jira ticket talks about problems with `QUORUM` reads from the `system_auth` keyspace, but why would we need to do that? Surely  `LOCAL_ONE` is sufficient, even if we have a `RF=ALL`, no?

**Comment by penberg** [comment]
Btw, I don't really see why `system_auth` keyspace should be any different from schema definitions in `system` keyspace, which are currently handled by special RPC calls. Perhaps we could just announce auth changes in the cluster similar to how we announce schema changes.

**Comment by slivne** [comment]
One reason maybe that having a conflict between schemas is bad but not as
bad as having a conflict between authentication

**Comment by glommer** [comment]
Can we get this moving again?

In the Cassandra bug tracker, they quote the following as a reason for this not working:

```
AllStrategy is actually bad to use for this. Auth caches stuff to make the queries faster, and if you do login with the default Cassandra user that uses QUORUM then you will rarely succeed at logging in if you have a ton of nodes.
There is a reason DataStax never contributed their EverywhereStrategy to OSS, it is very easy to screw yourself with it.
```
First, I don't get why this is a problem at all. Reads happen with LOCAL_ONE, except for this special Cassandra user. We could easily make it a requirement that this user is deleted before the second node joins. That seems a lot friendlier than having to update RF for auth for every node that joins manually.

In general, it seems to me that updates to the auth table should happen with CL=ALL. In the future, we should even use consensus to make sure everybody is aware of the update. Updates to auth are rare, and we can definitely push the complexity to the updates if need be.

**Comment by avikivity** [comment]
CL=ALL does not guarantee that all nodes will be updated (it just guarantees that you'll see a failure), and it makes it impossible to do auth changes when a node is down. We need a better method.

**Comment by glommer** [comment]
Bumping priority. This can cause availability issues cluster-wise as users forget this step when adding nodes.

**Comment by slivne** [comment]
@glommer - not clear what users are missing:
- do they forget to update the replication when they are adding a new dc ?
- docs have been updated on add node / add dc to include that 
https://docs.scylladb.com/operating-scylla/procedures/cluster-management/add_node_to_cluster/

We do not have a solution - using EveryWhereStrategy doesn't work 
- as all nodes need to be up and running todo auth changes. 

We can do something in the lines of https://issues.apache.org/jira/browse/CASSANDRA-14303
Thats is: specify a NetworkToplogyStrategy with replicaion_factor: 5 - that would mean that if any new dc is added it will be replicated to, if nodes are added then we will replicate up to 5.

**Comment by glommer** [comment]
Yes, people commonly forget things. And then they forgot they forgot and stuff just doesn't work.
A safer, more robust procedure is better than just a documentation step that people may miss.

There isn't much data in system_auth.

Why can't we automatically replicate everywhere, use quorum writes, and nodes that go up always auto-repair ?

**Comment by avikivity** [comment]
It's a good goal but paved with things that need to be considered.

1. QUORUM doesn't work if you lose internet access, so need LOCAL_QUORUM, which may not be up to date
2. Repair happens after the node is up, better a variant of bootstrap that streams from all available nodes
3. Or maybe available local nodes

A nice advantage of everywhere replication is that we can consider LOCAL_TWO reads (LOCAL_QUORUM can be a significant load in large clusters).

**Comment by tzach** [comment]
> Yes, people commonly forget things.

#4210 suggestion solve this, and not just for system_auth

**Comment by glommer** [comment]
I don't see how.

The problem with system auth doesn't manifest only when adding DCs.

It is more pervasive even when adding nodes to thr existing DC

**Comment by dorlaor** [comment]
Shouldn't we define something like RF=ALL for this and use CL=local_q ?

**Comment by tzach** [comment]
> The problem with system auth doesn't manifest only when adding DCs.
> It is more pervasive even when adding nodes to thr existing DC

This is what I do not understand. Why ALL nodes need the info, and not just ~5 nodes per DC?
If it's for HA, the later is good enough.

**Comment by glommer** [comment]
My understanding is that we want auth queries to be local since going across the network can add latency to to the queries.

When we originally discussed this, our auth cache was pretty bad. Maybe it is better now and we need to do this less often. Still, in practice we can have replicas that are completely pegged, and if they happen to be the ones that will hold auth data, the entire cluster will suffer.

I am not sure myself it there is any requirement *aside* from this.

**Comment by slivne** [comment]
We have implemented #4210 - which means that we can now specify replication of 3 (per dc) for all system tables and the users will not need to update replication. If a new DC will be added it will have 3 local replicas as well.

@glommer - would this be enough to close this ticket ?

**Comment by nyh** [comment]
I just got hit by this issue in #6440, where Alternator uses the `system_auth` table (issue #6440 is a test, but it also happens with Alternator itself) but it is created with RF=1 so if just one of the nodes is unavailable, pieces of this table are unavailable (and lost forever if one node is lost forever!).
I see large amounts of discussions about this, various warnings in scylla.yaml about system_auth replication, and also additional related issues such as #5878 and #6351. 
So isn't it about time we solve this issue?

**Comment by slivne** [comment]
@glommer ping on https://github.com/scylladb/scylla/issues/1796#issuecomment-573454671

Instead of everywhere I want to move to networktoplogystrategy with default replica_factor of 3.

**Comment by glommer** [comment]
I think that makes sense.
It would solve the issue with people having to change it manually.
Ever since we first opened the issue, we found and fixed many issues with scalability in the auth cache layer so we should be less dependent on CL=LOCAL_ONE

Go for it, but let's make this RF=3 per DC so we can keep the auth queries local

**Comment by geobeau** [comment]
Another argument in favor of using everywhere replication is that on large clusters the command "LIST ALL PERMISSIONS" will timeout (It is implemented as "SELECT role, resource, permission FROM system_auth.role_permissions;" with a consistency level of ONE) with RF 3 (because it will need to interact with a large number of nodes/shards). Adjusting the RF of the Network Topology automatically to the number of nodes might not be great as you can have consistencies issues between the change of RF and the effective end of the repair.

**Comment by eliransin** [comment]
/cc @vladzcloudius - I once uppon a time mentioned that moving to everywhere topology strategy isn't simple :slightly_smiling_face: 
Here is the discussion about it.

**Comment by vladzcloudius** [comment]
> /cc @vladzcloudius - I once uppon a time mentioned that moving to everywhere topology strategy isn't simple slightly_smiling_face Here is the discussion about it.

Ok. So the only "cons" I see is that QUORUM would need to read from many nodes.
But the same holds when we change the replication manually. So, this makes this "cons" invalid.
And node that if this does become an issue then one can always change to a different replication strategy with a different replication factor.

Note that we are talking about a default here.

**Comment by bhalevy** [comment]
IIUC when writing with EverywhereReplicationStrategy all nodes have to be up and this creates fragility.
If this is the case I wonder if we can support simple CL=QUORUM with RF=ALL, making sure we also read with CL=QUORUM (rather than CL=LOCAL_ONE).

**Comment by vladzcloudius** [comment]
> IIUC when writing with EverywhereReplicationStrategy all nodes have to be up and this creates fragility.

Only if you use CL=ALL.

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/issues/6440 was another duplicate of this issue - or at least be automatically solved when this issue is solved.

That issue was an Alternator test failure when running on a cluster with one node down:

When I start a three-node Alternator cluster and then kill one of them, almost all of the `test/alternator` tests work properly - as they should (Alternator never uses `CL=ALL`). However, one test fails:

```
$ pytest test_system_tables.py::test_fetch_from_system_tables
...
E           ClientError: An error occurred (Internal Server Error) when calling the Scan operation (reached max retries: 3): Internal server error: exceptions::unavailable_exception (Cannot achieve consistency level for cl LOCAL_ONE. Requires 1, alive 0)
```

The problem is that the "test_fetch_from_system_tables" test tries to read the system_auth table, and that has replication factor 1 by default.

**Comment by eliransin** [comment]
@roydahan here is a very long discussion about it.
The TL;DR is that it is undecisive and there are a lot of conflicting assumptions about everywhere replication strategy.
AFAIC if we document that RF must always be equal to the number of nodes in the cluster (which we do [here](https://docs.scylladb.com/stable/operating-scylla/security/authentication.html#prerequisites])), there is no real difference between Everywhere and any NetworkTopologyStrategy.
With that said, this is only true if Everywhere strategy really works as expected.
So I sugest the following:
1. Add some tests specifically to Everywhere strategy (Dtests?)
2. Try to run all tests with the system_auth changed to Everywhere strategy
3. Maybe some SCT runs with the changed procedure? (Do we have auth trageted SCT?)
From what I can see, the main fear here is from the devil we don't know so we keep hanging on to the devil we know :smiling_imp:

**Comment by roydahan** [comment]
I don't understand if it's a change we consider on doing or a change we've already done.

**Comment by mykaul** [comment]
> I don't understand if it's a change we consider on doing or a change we've already done.

In-between - we are not considering it, we are discouraging use of SimpleStrategy and are in the process of moving away from it across the board.

**Comment by bhalevy** [comment]
To make it clear, the system_distributed keyspace is still auto-created using SimpleStrategy.
See https://github.com/scylladb/scylladb/blob/54f174a1f4d184414be3aa3dc34889fcd468b729/db/system_distributed_keyspace.cc#L274-L278

**Comment by avikivity** [comment]
To reduce the scope, I propose we switch to raft group 0 (when raft is enabled) and no change when not.

The implementation is non-trivial since we must support upgrades.  We can implement these upgrades by running a full scan of the table and committing it to Raft. There's a race condition if someone writes something during the upgrade, but we'll have to live with it.

Those tables are small so no problem to materialize everything in memory.

**Comment by kbr-scylla** [comment]
There is one problem with `cdc_streams_descriptions_v2` and `cdc_generation_timestamps` in `system_distributed`, they are not exactly internal, they are used by CDC clients, so at least in terms of what clients see, they need to stay in `system_distributed`.

Internally, perhaps we could change them to virtual tables or something.

**Comment by mykaul** [comment]
@ptrsmrn , @kostja - shall we break this one into digestible pieces and see what can and cannot fit into 6.0?

**Comment by kostja** [comment]
I don't think moving Redis over to Raft is a priority, moving out of the scope of 6.0.

**Comment by nuivall** [comment]
point 1 from issue description was done, we can remove it from here

**Comment by kbr-scylla** [comment]
The user-facing CDC tables in `system_distributed` keyspace still exist and will be used by CDC connectors and clients.
We changed the internal tables.

**Comment by bhalevy** [comment]
Cc @paszkow

**Comment by bhalevy** [comment]
@avikivity should we deprecate EverywhereStrategy for user tables?
We don't intend to support them with tablets.

The following is possible today (2026.1-dev)
```
cqlsh> create KEYSPACE everywhere with replication = {'class': 'EverywhereStrategy'} and tablets = {'enabled': false};
```

Cc @tgrabiec
